### ISSUE TITLE
Lag on player join

### ISSUE THREAD

**Issue by Waffleman123s** [TeamUlysses/ulx#59]
I have installed the new version of ULX and ULIB and when ever a player joins the server spike lags for a few seconds and I need to know how to fix this asap!
I deleted all my addons and just left ULX and ULIB in and it still does it now when ever I delete ULX and ULIB the lag stops!!???

Please help me asap!

**Comment by sammyt291** [comment]
ULX / ULib version
OS version
Latest gmod version?
Errors in console?

**Comment by Waffleman123s** [comment]
0 ERRORS in console! We had the latest version of ULX and ULib and we create a fresh darkrp gamemode installed ULX and ULib FRESH and still Lags when people Connect!!!! FML

**Comment by Nayruden** [comment]
What OS? Are you using the latest version of GMod?

**Comment by sammaxis** [comment]
I believe I am experiencing the same issue as the OP.

It's a mild lag that only occurs on player join or leave. I have tried removing everything outside of ULX & ULIB as well as changing other settings.

Looking for some help.

**Comment by Mika32150** [comment]
Same problem here, please fix Ulib and ULX. Before we doesn't have this lag when someone join / disconnect.

**Comment by Mika32150** [comment]
For me i use Debian 8 with the last gmod update !

**Comment by davethestuff** [comment]
Also experiencing this problem. Think it started since the latest gmod update.

**Comment by Mika32150** [comment]
Someone work on the problem ?

**Comment by FlashModz** [comment]
i'm some problems.

https://github.com/TeamUlysses/ulx/issues/61

Debian 8

**Comment by Nayruden** [comment]
I'm unable to duplicate this problem, but I've never tested on Debian before (I use Ubuntu). Will try installing Debian now.

**Comment by Nayruden** [comment]
I still can't replicate this issue, and I don't see what Garry would have changed recently that would have resulted in this for Debian users. However, I took a wild stab at what the issue may be, so please try using the following slightly modified ULX and ULib versions and report back --

http://ulyssesmod.net/test_ulib.zip
http://ulyssesmod.net/test_ulx.zip

Note: due to the changes in these, there may be some errors. I tried to clean up any obvious problems, but this should work for this experiment.

**Comment by sammaxis** [comment]
I'm not using Debian. We're hosted on NFO and I'm experiencing the same issue.

**Comment by uRandomAlex** [comment]
@Nayruden You should create another GitHub branch and make changes there. I don't want to ruin all my ULib/ULX modifications to test this.

**Comment by uRandomAlex** [comment]
@Nayruden Have you tried to play on a live server, 10+ players (not bots)? Problems starts there. Every new joined player, that's on "Sending client info" stage (PlayerAuthed/PlayerInitialSpawn hooks), freezes server for a second.

**Comment by Nayruden** [comment]
I do not have the ability to test on such a server, @uRandomAlex.

**Comment by FlashModz** [comment]
@Nayruden actualy in the test

**Comment by FlashModz** [comment]
@Nayruden same problems on my Debian 8... with tes_ulx and ulib.

Solution ?

**Comment by Nayruden** [comment]
I need someone to volunteer their server for experimentation; otherwise, I'm just making wild guesses. I'm unable to do anything for the next week or two, as I am moving my household across the country.

**Comment by sammaxis** [comment]
I suppose I'd be able to do that.

http://steamcommunity.com/id/sam_maxis/

**Comment by Violat0r** [comment]
I can confirm we have this issue on our servers also, mainly on DarkRP. We are running on a Windows box. No errors, latest gmod, lastest ulib/ulx. When I use an addon like FProfiler and what not, its always ulx/ulib spiking in cpu when someone joins the game.

**Comment by FlashModz** [comment]
Solution ?

**Comment by Nayruden** [comment]
@FlashModz see https://github.com/TeamUlysses/ulx/issues/59#issuecomment-234315115.

**Comment by FlashModz** [comment]
I have a dedicated server I am willing to mount a dev server with all the access that will be delivered you add me on steam

FlashModz

**Comment by Nayruden** [comment]
Thank you, I will certainly take someone up on that offer. The larger problem is I am working out of a hotel at the moment. I'll be in touch when I am able to debug this. :)

**Comment by FlashModz** [comment]
if patched ?

**Comment by Nayruden** [comment]
@FlashModz see https://github.com/TeamUlysses/ulx/issues/59#issuecomment-234315115. I did get internet yesterday but have other tasks that take priority over ULX this weekend.

**Comment by Nayruden** [comment]
Trying to work out what could be causing this. Can one or more of you experiencing this please answer the following questions?

How many users do you have added in ULX? A rough estimation (10's, 100's, 1000's) is fine.
Does this happen every single time someone joins and leaves?
Is it worse with more people already in the server?
Any other trends you've noticed? (certain gamemodes, maps, etc)

**Comment by uRandomAlex** [comment]
@Nayruden 10 users, every single time someone joins and leaves, it's worse with more people on server.

**Comment by Nayruden** [comment]
@uRandomAlex, please run "lua_run PrintTable(ULib.ucl)" from you client and paste the results here or send it to me directly -- megiddo@ulyssesmod.net

**Comment by uRandomAlex** [comment]
@Nayruden Sent an e-mail.

**Comment by FlashModz** [comment]
Here are the 16 laggiest hooks found during the last 781 frames:
Identifier ------------------------------------------------------------ Impact Executions Mean t /f Max time 
Event "PlayerAuthed", hook "ULibAuth" 502.40% 1 247.408us 193226us 
in "addons/ulib/lua/ulib/server/ucl.lua" (lines 1093 to 1097),
not found in Workshop addons
Event "UCLChanged", hook "ulx.updateMotd.adminsChanged" 250.21% 2 246.434us 97313.6us
in "addons/ulx/lua/ulx/modules/sh/menus.lua" (lines 71 to 120),
not found in Workshop addons

10 players or more is very laggy my server
yes whenever a player joins but its not this remark that from 12 players
DarkRP - rp_rockford_french_v2b

Thanks :)

**Comment by FlashModz** [comment]
As mentioned if you need a server DEV I can provide you with a test for which I will participate with pleasure :)

**Comment by PolSpock** [comment]
What's up ?

**Comment by Nayruden** [comment]
I've carved out some time this weekend in order to look into this more. I need to get a game plan in place for how I'm going to debug on a live server, then I'll be in touch with those who volunteered server time.

**Comment by Nayruden** [comment]
I wasn't able to hail any of the volunteers, but someone please test that out and report back.

**Comment by FlashModz** [comment]
i'm currently testing

**Comment by Nayruden** [comment]
Can you hail me on Steam? http://steamcommunity.com/id/megiddo/

**Comment by FlashModz** [comment]
Yes you're already in my friends :)

**Comment by Nayruden** [comment]
What's your nickname on Steam

**Comment by Nayruden** [comment]
I'm not sure that the issue @FlashModz was having is related to this. For @FlashModz, his client lags when he joins his own server (as opposed to all currently connected clients lagging when anyone joins), but other clients join without lagging. I advised @FlashModz that he may have too many client addons installed.

I have still been unable to duplicate this issue anywhere. If anyone else is still having issues, feel free to throw your input in here. :)

**Comment by FlashModz** [comment]
if that does not bother you I warn you friend to invite you to when there lags if it reproduces

**Comment by Dayornight1999** [comment]
I have tried the potential fix on a live server and it didn't seem to fix the issue for me.

**Comment by Nayruden** [comment]
@Dayornight1999, just to make sure we are all on the same page, please describe what your issue is exactly. Please be very detailed.

**Comment by Dayornight1999** [comment]
I'll try to be as detailed as possible but there isn't much to talk about. Essentially players were starting to report lag spikes every now and then. I started to investigate by looking at net_graph 4 and came to the conclusion that every time a player joins/leaves it started to lag my server for a second. I deleted the addons that could have been the issue, but nothing fixed it. So I did some more research and landed on this page. I've tried the potential fix you posted a day ago, but this doesn't seem to fix it. It also breaks something with the ranks I think.

**Comment by Nayruden** [comment]
What happens with your ranks?

**Comment by Dayornight1999** [comment]
I'm not sure really, but this is an error I just received when someone left the server after refreshing the XGUI modules and refreshing the server data. 

Error in hook UCLChanged: addons/ulx/lua/ulx/xgui/bans.lua:516: [ULIB] Unauthed player
stack traceback:
[C]: in function 'query'
	addons/ulx/lua/ulx/xgui/bans.lua:516: in function 'fn'
	addons/ulib/lua/ulib/shared/hook.lua:110: in function <addons/ulib/lua/ulib/shared/hook.lua:93>
	[C]: in function '‌⁪‎⁭​‬⁯‪‭'
	lua/.dragondildos/rc1/../nul/5d0b655d3a21f56b78412174caf0834a6f1366dbd2dacc342be329434a82ee8f.lua:59: in function 'fn'
	addons/ulib/lua/ulib/client/cl_util.lua:22: in function 'func'
	lua/includes/extensions/net.lua:32: in function <lua/includes/extensions/net.lua:17>

[ERROR] addons/ulib/lua/ulib/client/cl_util.lua:22: 
  1. fn - [C]:-1
   2. func - addons/ulib/lua/ulib/client/cl_util.lua:22
    3. unknown - lua/includes/extensions/net.lua:32

**Comment by Dayornight1999** [comment]
From what I can tell it seems to be removing people from current groups then readding them again, but I might be wrong. It's a weird issue and I can't find a patern yet.

**Comment by Dayornight1999** [comment]
Also getting this when I switch maps: http://pastebin.com/vZAV33Z1

**Comment by Violat0r** [comment]
Well good news is that it did reduce the lag on the server overall but it still does lag on player join.

I am also getting a lot of errors. http://pastebin.com/JfvmbTRX

**Comment by Nayruden** [comment]
If you're getting errors, try removing and redownloading from Git(Hub); if you were using the zips I posted a while back, you will have conflicts without fully removing and reinstalling.

**Comment by Violat0r** [comment]
I am using the one from github not the link.

**Comment by Nayruden** [comment]
Did you ever use the zip? If so, you need to do a full reinstall.

**Comment by Violat0r** [comment]
Never used it and I always delete the old versions of addons.

**Comment by Dayornight1999** [comment]
It seems ULX or ULIB wasn't the issue after all. It seems to be the Achieve! addon from scriptfodder that was causing the lag spikes. If any of you have that on the server, I recommend you remove it for now and contact the addon creator just like I will do. Nayruden, thank you for trying to help us out and sorry for wasting some of your time investigating into this problem.

~Dayornight

**Comment by Nayruden** [comment]
Reverted until I can fix the bot issue

**Comment by Dayornight1999** [comment]
Removing Achieve! solved this issue for me, but maybe others are still having this issue despite not having Achieve! on their servers. We'll just have to wait for a response from them.

**Comment by uRandomAlex** [comment]
@Dayornight1999 I didn't even know about existance of that addon. Seems that the problem is with sending too much data to player or with too expensive calculations on player join in ULib/ULX. This issue is not affecting servers with only a few addons because they don't use CPU/network that much and therefore ULib/ULX can have this resources "right now", without need to wait for them.

**Comment by Dayornight1999** [comment]
Didn't the person some posts above us say he removed all his addons except for ULX/ULIB and he still get the lag spikes?

**Comment by Nayruden** [comment]
ULX performs no calculations on join and sends less than a kilobyte. That's
partly why I'm so confused why anyone is experiencing issues.

**Comment by Violat0r** [comment]
I honestly think it might be darkrp or an addon for darkrp that relies on ulx/ulib for perms. We have other servers with the same issue, all darkrp gamemode. Our other servers that are not darkrp such as TTT and what have you, do not have this issue at all.

**Comment by uRandomAlex** [comment]
@Nayruden May be ULX don't perform any calculations, but ULX does - UCL is quite heavy. I guess you should look into it.

**Comment by Nayruden** [comment]
I assume you meant ULib. My previous statement about ULX applies to ULib as
well.

**Comment by uRandomAlex** [comment]
@Nayruden Yeah, I've changed it but seems that GitHub doesn't send you new e-mail. Anyway, profiling with FProfiler on player join shows that UCL is most expensive.

**Comment by PolSpock** [comment]
same analysis as @uRandomAlex

**Comment by FlashModz** [comment]
@uRandomAlex idem

**Comment by SticklyMan** [comment]
Random question for all of you having issues, what is your value of 'ulx_showmotd' set to?

**Comment by FlashModz** [comment]
ULX MOTD has been desactivated

**Comment by Violat0r** [comment]
Disabled, using something else.

**Comment by Nayruden** [comment]
Please check if that makes any difference for you all.

**Comment by Nayruden** [comment]
Any luck with the changes from last week?

**Comment by sayaloli1** [comment]
i fixed it with changing to an older ulx / ulib version ( 2 versions older )

**Comment by Nayruden** [comment]
Haven't heard anything back, is it safe to close this issue as solved?

**Comment by Violat0r** [comment]
Nope still lags :(

**Comment by Nayruden** [comment]
I apologize this has been extremely frustrating for everyone (the developers included). It feels like we've chased a bunch of dead ends up to this point, and we still don't have any leads as to what the problem is. We'd like to see if there are any patterns among folks who are experiencing the lag -- please help us out by filling out this 2-3 minute survey https://goo.gl/forms/6y7jdpHpIfouvcHA3

**Comment by Dayornight1999** [comment]
I don't know if I'm the only one with this issue, but the bring command and teleport command seem to be broken. Ulx bring doesn't work (it keeps saying it can't find a place to put the target even though there's plenty of space) and ulx teleport teleports the player into you and not where you're aiming at. Just thought i'd mention it.

**Comment by Nayruden** [comment]
@Dayornight1999 I'd suspect a conflicting addon there, but that's off topic of this issue.

**Comment by uRandomAlex** [comment]
@Nayruden Have you tried comparing current ULX/ULib to older versions (2-3 from latest release)? People says that they don't have this issue. Try profiling both of them (old version without lag and new version) with FProfiler. I think there will be something that consumes many resources compared to old version.

**Comment by Violat0r** [comment]
We broke darkrp on the server and it got rid of the lag but of course that breaks other addons. I am fairly sure it is darkrp or a darkrp addon that uses ulx for perms and what have you. I will do more testing to try and narrow it down.

**Comment by Hazbelll** [comment]
A little off-topic here, but why is ULib a thing again? Does it stand for **U**seless **Lib**rary?

What Developers or Server Owners use it for anything other than ULX? Wouldn't it make more sense to have ULX use its own specifically tailored library and get rid of ULib?

**Comment by Nayruden** [comment]
That's more than just a little off-topic; if you'd like to discuss this
further on our forums (ulyssesmod.net), I'd be more than happy to share my
thoughts.

**Comment by Hazbelll** [comment]
Well it's a contributing factor to this issue, if not directly. Point stands, why make complicate things more than they need to be? Worse for performance, more annoying to debug and such.

**Comment by uRandomAlex** [comment]
@Hazbelll ULX4 has other things that hurts performance - for example, language it's written in, MoonScript.

**Comment by roboderpy** [comment]
Because of way how ULib UCL handles groups, it realoads config file **on any** change of groups, or player join. Because of slow *KeyValues parse* function, you getting this lag. If @Nayruden would mind about remaking parse functions, lags will be gone. Or, he can do ever more - change how UCL handles group changes.

**Comment by Nayruden** [comment]
Survey results indicate that is not the problem here, though your idea may
be worth exploring as well. Sorry I haven't had a chance to continue to dig
into this more lately; I've got another idea of what may be causing this
issue, just waiting for my schedule to clear up.

**Comment by sweepyoface** [comment]
I'm having this issue, and it's quite crippling.

**Comment by Nayruden** [comment]
I'm sorry it's taken us so long to address this issue to everyone's satisfaction; I believe I will have a free day this week, and if so, it will be dedicated to tackling this. Thanks for being patient and understanding.

**Comment by sweepyoface** [comment]
I'll be glad to offer you my server to test on if we haven't fixed it by that time.

**Comment by sweepyoface** [comment]
Could "[S_API FAIL] SteamAPI_Init() failed; SteamAPI_IsSteamRunning() failed." be contributing to it? I've tried to hours to fix that and I haven't been able to.

**Comment by Nayruden** [comment]
The survey results showed that most people having the issue are running DarkRP with 100+ addons. Even running DarkRP with lots of addons on my laptop, with eight players and other things bogging down my system, I still can't replicate this. Someone experiencing the lag please reach out to me on Steam today -- http://steamcommunity.com/id/megiddo/

**Comment by Nayruden** [comment]
Working with @Violat0r on this issue, it seems that the lag is caused by FAdmin or an interaction between FAdmin and ULX. Disabling FAdmin seems to address the issue.

I may try removing CAMI from ULib in an experimental branch to see if that solves the problem as well; unfortunately, we've had nothing but issues from DarkRP, FAdmin, and CAMI, and I don't have the patience to keep working around them.

**Comment by sweepyoface** [comment]
We seem to have solved the issue by simply renting a more powerful server.

**Comment by FPtje** [comment]
> I may try removing CAMI from ULib in an experimental branch to see if that solves the problem as well; unfortunately, we've had nothing but issues from DarkRP, FAdmin, and CAMI, and I don't have the patience to keep working around them.

I've heard nothing about this for quite some time now. Where is this sudden desparation coming from? I'll look into FAdmin MySQL performance, but that shit should all be asynchronous.

What module is being used, tmysql or mysqloo? Does the lag still happen when the other module is used?

**Comment by Violat0r** [comment]
tmysql, although I highly doubt it is sql that is causing the lag.

We want people to try disabling FAdmin on servers to verify our results yesterday and report back.

**Comment by FPtje** [comment]
The best way to debug lag issues is with a [profiler](https://github.com/FPtje/FProfiler). Just run your server as normal, preferably with all the addons. Run the profiler, serverside or clientside depending on whether the lag is happening serverside or clientside. Have someone join while the profiler is running. After the lag spike has happened, stop the profiler. Then observe the most expensive function calls (in FProfiler that should be the second tab). 

Screenshot the results, post them here.

FProfiler is very likely to give you the exact source of the lag. There will be **no** guessing.

**Comment by Violat0r** [comment]
We have done that already. I do not have the results anymore but it is very similar to [this past comment](https://github.com/TeamUlysses/ulx/issues/59#issuecomment-239668356)

FProfiler lags our purge server to death with 40 people on when it is on so we cannot use it sadly. I used https://steamcommunity.com/sharedfiles/filedetails/?id=682765484 and it also pointed to the same hook. We have come to the conclusion that the hook spikes with darkrp as our other servers do not lag and the survey results all point to darkrp. We believe it is FAdmin which is why we want people to test it out.

**Comment by FPtje** [comment]
The profiler of the past comment only mentions ULib. If like you say your profiler also refers to the same hook, then it's clearly ULib.

Also, you don't need to run FProfiler full time. It's most efficient when you start it right before someone joins and end it right after.

I can't see the data behind the survey, but it is most likely that most of the servers reporting lag spikes are running DarkRP, maybe even 100% of the server owners run DarkRP. Not necessarily because DarkRP causes the lag spikes, but simply because it's the most popular gamemode. Also, there are *many* DarkRP specific addons which could also be causing the lag spikes. Of course there's going to be a strong correlation between DarkRP and these lag spikes. This doesn't imply causation. 

The OP said that removing ULX and ULib solved the issue. @Nayruden is saying that removing FAdmin solves the issue. The profiler posted earlier on points at ULib. Unless **REAL** evidence is provided that FAdmin is directly causing a lag spike, there is **NO** reason to believe that it actually is. I don't trust any of you people's ability to judge whether a lag spike is "gone" when you remove an addon.

**Comment by Nayruden** [comment]
The profiler we ran yesterday pointed to something else entirely, and it wasn't a large enough impact to matter. True, DarkRP is a popular addon, but that's where the evidence points at the moment. I have done everything within my power to replicate the problem and have been unable to. It could be a common addon between the lagging servers, there's just no way to know for sure with what we have so far.

**Comment by sweepyoface** [comment]
I'd like to point out that the issue was still there with no addons at all (just not as bad) but vanished as soon as we upgraded the machine. That and the fact you said FProfiler lags you to death leads me to believe we're overthinking it and you just need better hardware.

**Comment by FPtje** [comment]
FProfiler can indeed be expensive when there are many players. That's because it logs every function call.

**Comment by mcNuggets1** [comment]
I only see people complaining about their 100+ addons shitty optimized pay to win server lagging.
I use ULX and ULib since I started my 4 servers, I never had any negative experiences with it, besides it disabling connect/disconnect messages sometimes and sometimes not.

**Comment by FlashModz** [comment]
My test Hooks : test players in join and spawn



Here are the 16 laggiest hooks found during the last 2788 frames:
 Identifier ------------------------------------------------------------ Impact  Executions  Mean t /f  Max time 
 Event "UCLChanged", hook "xgui_RefreshPlayerCmds"                       125.05% 1           4.827us    13456.3us
  in "addons/ulx/lua/ulx/xgui/commands.lua" (lines 289 to 357),
 Net receiver "urpc"                                                     45.27%  3           5.241us    14312.6us
  in "addons/ulib/lua/ulib/client/cl_util.lua" (lines 7 to 23),

Thanks all

**Comment by Mika32150** [comment]
Someone tried this ? https://scriptfodder.com/scripts/view/3389/ulx-leysql-web-bans-groupsmore

**Comment by Mika32150** [comment]
Doesn't work after 1 month, but why the user.txt keep "user" in ? I think user.txt need to keep only admins and VIP. I've more 8.000 lines in my user .txt adn i think it's because users.

**Comment by Nayruden** [comment]
"user" is not stored by default.

**Comment by mcNuggets1** [comment]
A smart way of saving data is creating a users/bans data folder and storing each ban as a new file in the specific folder.

**Comment by Hazbelll** [comment]
Or just do-away with this flat file database bullshit in .txt files and use SQLite and/or MySQL.

**Comment by mcNuggets1** [comment]
Adding SQLite support is a good solution, but dropping flatfile support completly is a terrible solution.

**Comment by Nayruden** [comment]
This is making the assumption that reading from files is causing lag, which
wasn't the case when we tested affected servers.

**Comment by Hazbelll** [comment]
Why on earth would you want flat file databases? SQLite is native to Garry's Mod and (in the long run) provides better performance. The only downside is portability (of ULX' data specifically), but you only need to use something like [SQLite Browser](https://github.com/sqlitebrowser/sqlitebrowser) to do the job anyway.

**Comment by mcNuggets1** [comment]
It is causing giant lag spikes, when the data file is pretty big.
For example my bans.txt has 2800 entries and when adding a new ban the servers freezes for 1 entire second.

**Comment by Nayruden** [comment]
mcNuggets, we've tested several thousand bans before and had no appreciable
delay. But, I'll test again to verify.

Hazbelll, ULX's saving of data far predates Garry's inclusion of SQLite
(which would have been what we used, had it been there). We have had no
convincing reason to move to it.

**Comment by Hazbelll** [comment]
@mcNuggets1 But why would anyone be *that* unwilling to learn to/bother using a SQL Browser, even a super simple GUI-based one such as the one I linked? I don't think the sort would even care about retaining/migrating data anyway, but just start again from scratch regardless.

@Nayruden It's been native to GMod for quite a long time now. You said you would've used it if it had been available at the time, what was your reason for that over choosing flat file and how come it has since changed?

**Comment by mcNuggets1** [comment]
@Hazbelll Its still more user friendly to keep flat files.
Optionally adding SQLite is okay for me, but not removing flat file support completly.

Besides I know no one who uses a SQL Browser besides me.

**Comment by Nayruden** [comment]
@Hazbelll, I already explained why we didn't move to SQLite.

@mcNuggets1, if we went with SQLite, we would remove flat-file support
(with one-time conversion from flat-file). It is difficult and error-prone
to support multiple formats. Not impossible, but difficult.

**Comment by Hazbelll** [comment]
@Nayruden No you didn't.
> (which would have been what we used, had it been there). We have had no
convincing reason to move to it.

That doesn't explain anything at all, other than you would have used it if it had been available originally, but now you won't because of no convincing reason to. **Why would you have picked SQLite if it had been available at the time, if there's no convincing reason to move to it now?**

**Comment by Nayruden** [comment]
@Hazbelll, ah, I misunderstood your question, my apologies. There's no
convincing reason to move to it, but it still would have been my first
choice if it had been available. E.G., I recognize there is benefit to
using it, but not enough benefit to rewrite what we currently have.

**Comment by mcNuggets1** [comment]
Every config setting, all groups etc. would break when we move to sqlite.
When the sql database gets corrupted, we loose everything.
We can't modify simple things without restarting the server etc.

**Comment by bigdogmat** [comment]
> Every config setting, all groups etc. would break when we move to sqlite.

You could easily write something to convert all the data. I'm sure if it was switched there'd be an auto converted put in place.

> We can't modify simple things without restarting the server or breaking shit etc.

You can modify the database while the server's up.

> When the sql database gets corrupted, we loose every settings.

I've never had my database get corrupted, though having a backup system wouldn't be bad.

**Comment by Hazbelll** [comment]
> Every config setting, all groups etc. would break when we move to sqlite.

Hence why the data would need a one-time conversion.
> We can't modify simple things without restarting the server or breaking shit etc.

Yes you can.
> When the sql database gets corrupted, we loose every settings.

Same applies with whatever storage method is used and why regular backups are incredibly important. Don't tell me you don't already do that?
> Too many negative aspects.

If you have no idea what you're doing and no incentive to learn, yes. But then you shouldn't be running a server if you don't know what you're doing.

**Comment by mcNuggets1** [comment]
I do backups of course, but not regularly.
Still things could get lost. Even with backups. (I won't do backups every 10 minutes)

I am just not a friend of saving to the SQLite database.

I like things uncomplicated. 
I myself am capable of scripting in GLua, but I ain't professional.
I also save some things on the SQLite database, but I don't want to save ULX related things onto it.
I am trying my best to keep a great server, so don't tell me that shit.
I just don't see any point to port ULX to SQLite besides a slight performance impact.
It is more a hassle than a liberation.

**Comment by sweepyoface** [comment]
> (I won't do backups every 10 minutes)
Why not?

**Comment by mcNuggets1** [comment]
I really don't feel the way saving a 30-40mb sqlite file every 10 minutes.

**Comment by sweepyoface** [comment]
> I really don't feel the way saving a 30-40mb sqlite file every 10 minutes.

Let me introduce you to [rdiff-backup](http://www.nongnu.org/rdiff-backup/).

**Comment by Hazbelll** [comment]
> There's no
convincing reason to move to i

@Nayruden How about future-proofing? As in for example, when/if that ban table grows to a humongous size (Some servers last years with their data retained).

[The GMod Wiki also states SQLIte is the fastest (At least natively) storage method.](https://wiki.garrysmod.com/page/Category:sql)
![Image of Wiki SQLite Statement](https://cdn.discordapp.com/attachments/193035695722463233/307158953245016064/unknown.png)

Most larger communities end up ditching ULX' built-in solution for an external one (Such as SourceBans) or even ditching ULX altogether for their own admin mod due to performance and other limitations (Such as hosting multiple servers wanting to share data between).

Bear that in mind when it comes to why maybe such issues don't appear so prominent, because everyone just tip-toes around it, due to the lack of official support from ULX.

**Comment by mcNuggets1** [comment]
I already said, ULX should not save to one big file.

**Comment by Nayruden** [comment]
@Hazbelll, SQLite does not help for cross-server data saving. Additionally,
there are enough differences between SQLite and other SQL flavors (such as
MySQL) that SQLite support does not lend itself all that well to MySQL
support. Nor is there native support for other SQL flavors in GMod natively
anyways; the average user is not able to install an SQL module.

**Comment by sweepyoface** [comment]
The "average user" doesn't need to worry about using a database anyways. Why not support both?

**Comment by Hazbelll** [comment]
@Nayruden 
> SQLite does not help for cross-server data saving.

I already stated that, soon after posting. Read the edit.
> Additionally, there are enough differences between SQLite and other SQL flavors (such as
MySQL) that SQLite support does not lend itself all that well to MySQL
support. Nor is there native support for other SQL flavors in GMod natively
anyways

If you don't want/can't be bothered to support two flavours of SQL, then just use an abstraction layer, like @FPtje and his [MySQLite Module](https://github.com/FPtje/MySQLite).
> the average user is not able to install an SQL module.

The "average user" lets their TCAdmin do that for them, or as @sweepyoface outlined.

**Comment by Nayruden** [comment]
@Hazbelll, not sure how familiar you are with SQL, but that abstraction
module would only work in simple use cases. As for not supporting both, we
are back to my earlier point of not supporting more than one method.

**Comment by Hazbelll** [comment]
@Nayruden I wouldn't know. I've never used it myself (Not needed such). Address that with @FPtje if that's the case.

But *why* won't you support more than one method? *Why* is any suggestion given turned down for some arbitrary reason? [Even a simple cvar to natively set ULX to log into Source's own logs or its own was turned down](https://github.com/TeamUlysses/ulx/issues/5).

You fear doing things that "the average user" would have no use for, but then totally ignore any suggestions or feedback given by those who actually come forward with such.

**Comment by mcNuggets1** [comment]
It is still a piece of a work.

If you need it so much, do it yourself and send a pull-request.

**Comment by Hazbelll** [comment]
> If you need it so much, do it yourself and send a pull-request.

Already said I anticipate on ditching ULX entirely for my own thing. I understand their team might be small or tied up with other matters and I respect that, but the lack of interest in anything put forward to them is just disheartening. I'm only here now because I got email alerts today for **an almost one year old issue**.

**Comment by Nayruden** [comment]
@Hazbelll, I agree with @mcNuggets1 on this. We are trying to manage
several responsibilities to the best of our ability:

1. We serve all our users, not just you. Just because you *feel* a decision
is arbitrary, I promise you it's not. We carefully consider what our users
tell us; this is why we are having this discussion at all.
2. We have limited time for development and bug-fixes; thus why I'm very
hesitant to support multiple formats, as we cannot maintain the quality we
would like in that case.

I explained why we turned down the cvar, as it has minimal benefit to our
users (mostly, to you). It is not a feature, it is just a way for you to
save a few hundred kilobytes. We have many requests for such things, and
while they seem alright individually, they turn the project into a monster
taken together. As an example, I got a request to make an API for
adding/removing items from ULib's tool whitelisting. I turned it down
because table.insert and table.remove do the job nearly as well as any API
would, without adding bulk for our users.

**Comment by Nayruden** [comment]
@Hazbelll, this issue should probably be closed, but I was accepting
feedback to see if anyone else had this problem. The servers I checked out
were not lagging due to ULX, but I want to make sure that I'm not missing
something that might be causing this for other users. If you had read the
comments, you would know that. Overall, no one is forcing you to use our
free product -- please feel free to use something else. I am saddened that
you feel you weren't listened to, but I offer no apologies.

**Comment by mcNuggets1** [comment]
Still, reading and writing to huge files IS causing lag.
Changing users.txt to a folder containing steamids and some 3 line group texts/privileges would be enough to resolve the lag, caused by the users.txt file.

I don't know how much work that would be, but I guess, when you decline to change these specific things , we have to accept that.

**Comment by Nayruden** [comment]
@mcNuggets1, again, we have tested huge files and noticed no lag. I will
test again to verify.

**Comment by mcNuggets1** [comment]
I noticed lag on dedicated servers. But not on listen servers.

**Comment by Nayruden** [comment]
I hardly ever test on a listen server. Can you send me your server files?

**Comment by Hazbelll** [comment]
> We serve all our users, not just you.

Seems you've not bothered to read half of what I said earlier then. As for myself, again. Already said I'm ditching ULX because yes, it'll never be suitable for my specific needs, hence why I take matters in-house there.
> We have limited time for development and bug-fixes; thus why I'm very
hesitant to support multiple formats, as we cannot maintain the quality we
would like in that case.

And like I already said, I respect that, except it seems **nothing** is do-able. If nobody has the time anymore, maybe it's time to pass on the torch?
> I explained why we turned down the cvar, as it has minimal benefit to our
users (mostly, to you). It is not a feature, it is just a way for you to
save a few hundred kilobytes.

It was an example, but if we dig into there: I know of other server operators that make use of the native Source logs, because (as I outlined in that issue) it supports logging of events outside the realm of Lua, such as engine errors. It *is* a feature request that maybe some of us **do not want redundant data**. It's such a god damn simple thing to implement and support.
> they turn the project into a monster taken together. As an example, I got a request to make an API for
adding/removing items from ULib

Stop right there. ULib itself is a monstrosity of all monsters. If you want to cut down on things to support, removing ULib's a good start. Maybe once it served a good purpose, but these days I've never encountered *anyone* using it, unless it's for ULX as it depends upon it or ULX-specific modules.
> Overall, no one is forcing you to use our free product -- please feel free to use something else.

Again, already stated I anticipate on moving away from ULX already. Win - Win. Maybe for anyone else who decides to do the same, or even make their own to serve the community, [maybe you and your little chums of Ulysses shouldn't be so bigoted about it](https://forums.ulyssesmod.net/index.php/topic,3655).

**Comment by sammaxis** [comment]
We still do have issues with lag fro. ULX on our server. We haven't upgraded to the latest version though. I don't believe that's the issue however. We've had the issue for over a year now. It only happens with ULX. If we uninstall all addons except ULX, we get lag when players join. It's not as bad as it used to be. We've honestly just learned to deal with it. Even if it is ULX which I'm not 100% sure it is, ULX and ULIB have been amazing, high quality free addons. I'm proud to run a server for free no pay-to-win with no hidden agendas and I'm happy ULX and ULIB similarly are great free products.
P.S. This thread got derailed though not much Nayruden could do about it. Though I know he's followed the issue since the beginning.

**Comment by Nayruden** [comment]
@Hazbelll, your reply once again seems to be all about you, not the
community. ULib does have many other projects utilizing it, even if you
aren't personally aware of them or using them. The divide between ULX and
ULib has worked well for the community. Hopefully, we can agree to disagree
on these matters. If not, I am still happy to have served you well up to
this point. I have been nothing but respectful towards you, and I am
saddened to not have that returned.

**Comment by mcNuggets1** [comment]
https://www.dropbox.com/sh/t447m8e7mx7w2c7/AABrgSV-SLrR2T5fq5jYfxnJa?dl=1

**Comment by Hazbelll** [comment]
@Nayruden 
> your reply once again seems to be all about you, not the community.

Earlier you claimed the "average user" was incapable of achieving something so simple as installing a SQL module, which feels like you're essentially saying the community is completely inept.
> ULib does have many other projects utilizing it

Oh then by all means point me to them then as I'm actually genuinely interested. Yes, myself and everyone else I know that works or has worked with ULib hasn't encountered such projects.
> I have been nothing but respectful towards you, and I am saddened to not have that returned.

Oh no, don't get me wrong. I have respect for you, but I do question the work ethics towards ULX sometimes and how people (not just myself) are addressed about some things.

That's your choice though, seems anyone else jumps on a bandwagon of either making other admin mods in spite of yours (Don't approve of that, but nor do I approve of the spite back. Two wrongs don't make a right) or just making their own for their own specific needs.

**Comment by sammaxis** [comment]
@Nayruden I would agree with the statement that the average user wouldn't understand how to install SQL. That doesn't mean they couldn't learn and it certainly doesn't mean they are inept. However you really are derailing the thread into your personal argument rather than this thread being used to actually address issues.

**Comment by Nayruden** [comment]
@sammaxis agreed on all points

**Comment by Mika32150** [comment]
So why i have user players in users.txt ?? http://prntscr.com/f1jir5

**Comment by sweepyoface** [comment]
Yeah, DB discussion aside, storing all users, bans, or whatever in a single file is completely unnecessary and it should be split up, kind of like how ItemStore does it.

**Comment by Radray** [comment]
My server was using the latest ULX at the time (around September last year) and we had roughly 3200 bans (550kb or so filesize). Whenever we banned someone there would be roughly a half a second lagspike.

Our solution was to make our own system for managing bans, but I do want to mention that all the effort that has gone into ULX and ULib is very much appreciated as my server used it for well over two years.

If you are absolutely unable to reproduce the lagspikes people are reporting relating to bans, I could possibly check if I have a backup still of the ulib datafiles that were doing this.

Regarding SQL/sqlite, I have to agree that it isn't for the average user. I've appreciated ULX saving data in flatfiles where it is easy accessible, however in terms of performance, SQL is superior to flatfile storage.

I am not entirely sure why ULib uses banid and from what I remember depends on banned_user.cfg when the CheckPassword hook can be used to prevent banned users from joining. I don't know if this has a performance impact though (it does seem to exec the banned_user.cfg every time someone is banned)

As for saving to a large flatfile, one could limit this to only once at startup, where you could load up the flatfile as usual, apply changes that were written to a separate file from before the restart, make a new save of the flatfile with updated values and delete the contents of the separate file. The server would then have an empty file to append lines to for desired changes on the next restart.

**Comment by sweepyoface** [comment]
You don't have to reproduce the lag to understand why rewriting a huge file every time someone gets banned would be a problem.

**Comment by Nayruden** [comment]
It's not a problem if it doesn't cause lag or any other negative side
effects. In general, If I'm unable to reproduce an issue, I'm unable to
determine the cause and therefore determine how to fix it.

**Comment by sweepyoface** [comment]
It certainly does, even if you have hardware fast enough so that it's barely noticable, go ahead and look at the net graph when you ban someone with 3000 other bans on the server.

I've already stated that the way to fix it is to split them into their own seperate files, so you aren't writing 3MB of data for every ban. This is ridiculously simple.

**Comment by Nayruden** [comment]
I am not disagreeing with you. I will test it again (from my laptop VM,
which is not fast) and share the results.

**Comment by Mika32150** [comment]
And for users in users.txt do you have an idea @Nayruden

**Comment by Nayruden** [comment]
@Mika32150, I'd have to look into it further, but my first thought would be
that you have an addon that's pushing the users into the file. That is not
default behavior for ULib. Are you using FAdmin?

**Comment by Mika32150** [comment]
Yes i've 3 darkrp servers ! I think user are here because i use : https://www.gmodstore.com/scripts/view/565/prometheus-a-gmod-donation-system

When someone stop to buy vip prometheus make him user and keep him in user.txt! 
What do you think ?

**Comment by Nayruden** [comment]
@mcNuggets1, can you re-send me your files, please? I wasn't able to download them before.

**Comment by Mika32150** [comment]
@Nayruden ?

**Comment by mcNuggets1** [comment]
https://www.dropbox.com/sh/te1okl94rhx6jnd/AAAc0rqwTef99wYTWt1AaFWva?dl=0

**Comment by Mika32150** [comment]
@Nayruden : i've found a possible fix for that issue.

Lot of server owner have the Prometheus addon, it's a donation system for give rank, money etc.. 
Link : https://www.gmodstore.com/scripts/view/565/prometheus-a-gmod-donation-system

I've check my user.txt and i've over 8000 lines into. The problem is, my user.txt keep player with "user" group, it doesn't erase it.

I think if the user.txt has too many entries, it make server lag when someone join because he need to check if player has a group in these list.

I've contacted the creator of Prometheus and he reply that : 

"Well that isn't a problem with prometheus though, it uses built-in functions from ULib to add/remove ranks, so if it doesn't remove the user from there, it isn't Prometheus fault, itš not like it adds/removes them from the file manually. Looking at their code, I can see it's because ULX/ULib doesn't check if the "addUser" function that is used to set ranks is trying to set it to default rank, it just sets it and doesn't care to what. The same "issue" would happen if you used the "addUser" thing in the ULX menu, it would also leave it in the file. They need ot implement a check in the "addUser" function that checks if the group they are being set to is default, remove the user fro mthe file then, instead of just changing the rank to user."

Can you fix this problem ?

**Comment by Mika32150** [comment]
18 days.. Nice support

**Comment by Heyter** [comment]
:D ulx shit.
use serverguard.

**Comment by FarukGamer** [comment]
ServerGuard bullshit, Open Source FTW!

**Comment by sweepyoface** [comment]
The fact that the issue of storing everything in one file hasn't yet been addressed is really concerning. I fear that common technical knowledge is being dismissed in favor of this issue being 'not reproducible'.

It's unfortunate that a lot of people including myself are forced move off of ULX because of such a large but simple problem. You have a responsibility to this project, and acting this way this says a lot.

**Comment by Hazbelll** [comment]
They can't be bothered, it's as simple as that.

**Comment by Nayruden** [comment]
@sweepyoface there's no need to over-optimize when the current solution does not create any problems. We've been around and around on this, please review the discussion.

@Hazbelll yes, our constant updates to a long-term project (over ten years of history and support) clearly demonstrates our lack of concern.

We still have yet to see lag on player join in the wild... if you have a server you believe is affected, please come forward.

**Comment by sweepyoface** [comment]
I have never seen someone exercise such an absence of intellectual clarity. As I mentioned before, you don't need to reproduce the lag to understand why you shouldn't be rewriting a 3MB+ file constantly.

How is this over optimizing when many people have complained about lag due to this specific issue, how the hell can you say this doesn't create problems?!

**Comment by Codingale** [comment]
Personally we don't have lag on join (least from ULX, our issue is GCompute) but we do however have issues with banning and adding people to ranks with a lot of bans. We used to have 5000~ and the server would freeze for 5 seconds. 

Removing all bans solved this I assume it's because of the function that sanitizes it as a strange table. We have a `bans,txt` but **not** the original` bannedusers.cfg ` if you need.

**Comment by Nayruden** [comment]
@Codingale please send it my way -- megiddo@ulyssesmod.net. Thanks!

**Comment by Codingale** [comment]
You should have it.


________________________________

**Comment by Nayruden** [comment]
Thanks, @Codingale. I've created an issue for it: https://github.com/TeamUlysses/ulib/issues/37

**Comment by Hazbelll** [comment]
![Hypocrisy 1](https://cdn.discordapp.com/attachments/193035695722463233/355366753104297984/unknown.png)

![Hypocrisy 2](https://cdn.discordapp.com/attachments/193035695722463233/355366821064474624/unknown.png)

Step in the right direction regardless.

**Comment by zpetty33** [comment]
Issue should be fixed in latest version of ULib.
